node/src/store/sled_store.rs

```diff
@@ -0,0 +1,497 @@
+use std::convert::Infallible;
+use std::io;
+use std::ops::Deref;
+use std::path::Path;
+
+use async_trait::async_trait;
+use celestia_types::hash::Hash;
+use celestia_types::ExtendedHeader;
+use directories::ProjectDirs;
+use sled::transaction::{ConflictableTransactionError, TransactionError};
+use sled::{Db, Error as SledError, Transactional, Tree};
+use tempdir::TempDir;
+use tendermint_proto::Protobuf;
+use tracing::debug;
+
+use crate::store::Store;
+use crate::store::{Result, StoreError};
+
+const HEAD_HEIGHT_KEY: &[u8] = b"KEY.HEAD_HEIGHT";
+const HASH_TREE_ID: &[u8] = b"HASH";
+const HEIGHT_TO_HASH_TREE_ID: &[u8] = b"HEIGHT";
+
+#[derive(Debug)]
+pub struct SledStore {
+    db: Db,
+    headers: Tree,
+    height_to_hash: Tree,
+}
+
+impl SledStore {
+    pub fn new(network_id: &str) -> sled::Result<Self> {
+        let Some(project_dirs) = ProjectDirs::from("co", "eiger", "celestia") else {
+            return Err(sled::Error::Io(io::Error::new(
+                io::ErrorKind::Other,
+                "Failed to get cache directory",
+            )));
+        };
+        let mut db_path = project_dirs.cache_dir().to_owned();
+        db_path.push(network_id);
+
+        Self::new_in_path(db_path)
+    }
+
+    pub fn new_temp() -> sled::Result<Self> {
+        let tmp_path = TempDir::new("celestia")?.into_path();
```

Comment:
This line and in general all the operations done in `SledStore` are IO operations, which means they are blocking and they will block executor. They need to be wrapped with `spawn_blocking` or `block_in_place`.

`spawn_blocking` is more preferred because if will use a new thread. With `block_in_place` we will restrict user to use only multithreaded Tokio executor.

You will need to have something like the following because `spawn_blocking` needs `'static`:

```rust
struct SledStore {
    inner: Arc<Inner>,
}

struct Inner {
    db: Db,
    headers: Tree,
    height_to_hash: Tree,
}
```

and you will use this like this:

```rust
    pub async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader> {
        let inner = self.inner.clone();
        
        spawn_blocking(move || {
            let hash = read_hash_by_db_key(&inner.height_to_hash, &height_to_key(height))?;
            get_by_hash(&inner.headers, &hash)
        }).await
    }
```

Reply:
Thanks for pointing that out, there's even [note](https://github.com/spacejam/sled#interaction-with-async) on the repo about async :sweat: 
Switched to `spawn_blocking`, reorganising code a bit to avoid spawning threads multiple times per single call.

Reply:
Sorry I didn't read their note before. To my understanding all read and write are cached and it is up to a point async friendly. I read some of their code just to make sure.

Based on this facts we will can not trust that blocking will never happen, but because it is going to a rare event (I think), we can switch to `block_in_place` instead. I'm trying to figure out how rare will be in our case. For example when another client tries to sync from us.

Reply:
Oh, not sweat, I appreciate the comment and agree we should have that fixed. 
In my testing I didn't notice blocking since default cache size of 1GB, which, by the looks of it, translates to roughly 10k headers cached.
Shame there doesn't seem to be a way to get a hint whether header we want is cached, but would it be worthwhile to do some kind of heuristic `block_in_place/spawn_blocking` decision? E.g. we hope last N headers are cached and block in place there, but do spawn_blocking for the rest to accomodate full syncs.
I'll switch to `block_in_place` for now.

Reply:
Per call, leaving it with `block_in_place`. Fixed remaining possible places where we could block.